File: src/test/java/tech/sirwellington/alchemy/generator/CollectionGeneratorsTest.java
/*
 * Copyright © 2026. Sir Wellington.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 *
 * You may obtain a copy of the License at
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package tech.sirwellington.alchemy.generator;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.RepeatedTest;
import org.junit.jupiter.api.Test;
import org.mockito.AdditionalAnswers;
import org.mockito.Mock;
import org.mockito.Mockito;

import java.util.ArrayList;
import java.util.UUID;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.*;
import static org.mockito.Mockito.*;
import static tech.sirwellington.alchemy.generator.AlchemyGenerator.one;
import static tech.sirwellington.alchemy.generator.NumberGenerators.integers;
import static tech.sirwellington.alchemy.generator.NumberGenerators.positiveIntegers;
import static tech.sirwellington.alchemy.generator.StringGenerators.*;
import static tech.sirwellington.alchemy.generator.Throwables.assertThrows;

/**
 * Tests for {@link CollectionGenerators}.
 *
 * @author SirWellington
 */
@DisplayName("Collection Generators")
final class CollectionGeneratorsTest extends BaseGeneratorTest {

    @Mock
    private AlchemyGenerator<Object> mockGenerator;

    @DisplayName("cannot be instantiated")
    @Test
    void testCannotInstantiate() throws Exception {
        assertThrows(
            () -> CollectionGenerators.class.getDeclaredConstructor().newInstance()
        ).isInstanceOf(IllegalAccessException.class);
    }

    @DisplayName("listOf(AlchemyGenerator) produces non-empty list")
    @RepeatedTest(DEFAULT_ITERATIONS)
    void testListOf_AlchemyGenerator() {
        var expectedValue = new Object();
        when(mockGenerator.get()).thenReturn(expectedValue);

        var result = CollectionGenerators.listOf(mockGenerator);

        assertThat(result, notNullValue());
        assertThat(result.isEmpty(), is(false));

        result.forEach(item -> assertThat(item, is(equalTo(expectedValue))));
    }

    @DisplayName("listOf(AlchemyGenerator, int) produces list of specified size")
    @RepeatedTest(DEFAULT_ITERATIONS)
    void testListOf_AlchemyGenerator_int() {
        // Given
        AlchemyGenerator<String> valueGenerator = mock(
            AdditionalAnswers.delegatesTo(alphabeticStrings())
        );
        var size = one(integers(10, 100));
        // When
        var result = CollectionGenerators.listOf(valueGenerator, size);

        // Then
        assertThat(result, notNullValue());
        assertThat(result.size(), is(size));
        verify(valueGenerator, times(size)).get();
    }

    @DisplayName("mapOf(String, String) produces map with UUID keys")
    @RepeatedTest(DEFAULT_ITERATIONS)
    void testMapOfWithInt() {
        var expectedValue = "fixed-string-for-test";
        var valueGenerator = AlchemyGenerator.of(() -> expectedValue);
        var size = 10;

        var result = CollectionGenerators.mapOf(
            StringGenerators.UUIDS,
            valueGenerator,
            size
        );

        assertThat(result, notNullValue());
        assertThat(result.size(), is(size));

        for (var entry : result.entrySet()) {
            var uuid = UUID.fromString(entry.getKey());
            assertThat(uuid, notNullValue());
            assertThat(entry.getValue(), is(expectedValue));
        }
    }

    @DisplayName("mapOf(AlchemyGenerator, AlchemyGenerator) produces non-empty map")
    @RepeatedTest(DEFAULT_ITERATIONS)
    void testConvenienceMapOf() {
        // Given
        AlchemyGenerator<String> keyGen = mock(AdditionalAnswers.delegatesTo(uuids()));
        AlchemyGenerator<String> valueGen = mock(AdditionalAnswers.delegatesTo(strings()));

        // When
        var result = CollectionGenerators.mapOf(keyGen, valueGen);

        // Then
        assertThat(result, notNullValue());
        assertThat(result.isEmpty(), is(false));

        // Given
        var size = result.size();
        verify(keyGen, times(size)).get();
        verify(valueGen, times(size)).get();
    }

    @DisplayName("fromList(Iterable) returns generator selecting from list")
    @Test
    void testFromList() {
        // Given
        var list = new ArrayList<String>();
        var size = RANDOM.nextInt(1, 100);

        for (int i = 0; i < size; ++i) {
            var value = Integer.toHexString(RANDOM.nextInt());
            list.add(value);
        }

        // When
        var generator = CollectionGenerators.fromList(list);
        assertThat(generator, notNullValue());

        // Then
        repeatBlock(DEFAULT_ITERATIONS, () -> {
            var value = generator.get();
            assertThat(
                "Value '" + value + "' not in source list",
                list.contains(value), is(true)
            );
        });
    }

    @DisplayName("listOf handles edge cases")
    @RepeatedTest(DEFAULT_ITERATIONS)
    void testListOfEdgeCases() {
        // Given
        var badSize = -5;
        var generator = uuids();

        // When, Then
        assertThrows(() -> CollectionGenerators.listOf(generator, badSize))
            .isInstanceOf(IllegalArgumentException.class)
            .hasSomeMessage();

        // Then
        var result = CollectionGenerators.listOf(generator, 0);
        assertThat(result, notNullValue());
        assertThat(result, is(empty()));
    }

    @DisplayName("mapGeneratorOf returns generator from parameters")
    @RepeatedTest(DEFAULT_ITERATIONS)
    void testMapGeneratorOf() {
        // Given
        AlchemyGenerator<String> keys = mock(AdditionalAnswers.delegatesTo(uuids()));
        AlchemyGenerator<Integer> values = mock(AdditionalAnswers.delegatesTo(positiveIntegers()));
        var size = one(integers(10, 100));

        // When
        var generator = CollectionGenerators.mapGeneratorOf(keys, values, size);

        // Then
        assertThat(generator, notNullValue());
        var map = generator.get();
        for (var entry : map.entrySet()) {
            assertThat(UUID.fromString(entry.getKey()), notNullValue());
            assertThat(entry.getValue(), notNullValue());
        }
        verify(keys, times(size)).get();
        verify(values, times(size)).get();
    }

    @DisplayName("mapGeneratorOf - edge cases")
    @Test
    void testMapGenerator_EdgeCase() {
        assertThrows(
            () -> CollectionGenerators.mapGeneratorOf(
                null,
                strings(),
                10
            )
        ).isIllegalArgumentException();
        assertThrows(
            () -> CollectionGenerators.mapGeneratorOf(
                strings(),
                null,
                10
            )
        ).isIllegalArgumentException();
        assertThrows(
            () -> CollectionGenerators.mapGeneratorOf(
                strings(),
                strings(),
                -10
            )
        ).isIllegalArgumentException();

    }
}

File: src/test/java/tech/sirwellington/alchemy/generator/CollectionGeneratorsTest.kt
/*
 * Copyright © 2019. Sir Wellington.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 *
 * You may obtain a copy of the License at
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package tech.sirwellington.alchemy.generator

import com.nhaarman.mockito_kotlin.whenever
import org.apache.commons.lang3.RandomUtils
import org.hamcrest.Matchers.`is`
import org.hamcrest.Matchers.empty
import org.hamcrest.Matchers.notNullValue
import org.junit.Assert.assertThat
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.Mockito.atLeast
import org.mockito.Mockito.mock
import org.mockito.Mockito.verify
import org.mockito.junit.MockitoJUnitRunner
import tech.sirwellington.alchemy.generator.AlchemyGenerator.Get.one
import tech.sirwellington.alchemy.generator.NumberGenerators.Companion.integers
import tech.sirwellington.alchemy.generator.NumberGenerators.Companion.negativeIntegers
import tech.sirwellington.alchemy.generator.NumberGenerators.Companion.positiveIntegers
import tech.sirwellington.alchemy.generator.StringGenerators.Companion.hexadecimalString
import tech.sirwellington.alchemy.generator.StringGenerators.Companion.strings
import tech.sirwellington.alchemy.generator.Throwables.assertThrows
import java.util.ArrayList
import java.util.UUID

/**

 * @author SirWellington
 */
@RunWith(MockitoJUnitRunner::class)
class CollectionGeneratorsTest
{

    private var iterations: Int = 0

    @Before
    fun setUp()
    {
        iterations = RandomUtils.nextInt(500, 5000)
    }

    @Test
    fun testCannotInstantiate()
    {
        println("testCannotInstantiate")

        assertThrows { CollectionGenerators() }
                .isInstanceOf(IllegalAccessException::class.java)

        assertThrows { CollectionGenerators::class.java.newInstance() }
                .isInstanceOf(IllegalAccessException::class.java)
    }

    @Test
    fun testListOf_AlchemyGenerator()
    {
        println("testListOf_AlchemyGenerator")

        val value = Any()
        val generator = mock(AlchemyGenerator::class.java)
        whenever(generator.get()).thenReturn(value)

        val result = CollectionGenerators.listOf(generator)
        assertThat(result.isEmpty(), `is`(false))
        result.forEach { assertThat(it, `is`(value)) }
    }

    @Test
    fun testListOf_AlchemyGenerator_int()
    {
        println("testListOf_AlchemyGenerator_int")

        val value = Any()
        val size = 50
        val generator = mock(AlchemyGenerator::class.java)
        whenever(generator.get()).thenReturn(value)

        val result = CollectionGenerators.listOf(generator, 50)

        assertThat(result, notNullValue())
        assertThat(result.size, `is`(size))

        result.forEach { assertThat(it, `is`(value)) }
    }

    @Test
    fun testMapOfWithInt()
    {
        println("testMapOfWithInt")

        val string = strings(50).get()
        val valueGenerator = AlchemyGenerator { string }
        val size = integers(5, 100).get()

        val result = CollectionGenerators.mapOf<String, String>(StringGenerators.uuids, valueGenerator, size)
        assertThat(result, notNullValue())
        assertThat(result.size, `is`(size))

        for ((key, value) in result)
        {
            UUID.fromString(key)
            assertThat(value, `is`(string))
        }

    }

    @Test
    fun testConvenienceMapOf()
    {
        println("testConvenienceMapOf")

        val keyGenerator = mock(AlchemyGenerator::class.java)
        val valueGenerator = mock(AlchemyGenerator::class.java)

        whenever(keyGenerator.get())
                .thenAnswer { one(strings()) }

        whenever(valueGenerator.get())
                .thenAnswer { one(positiveIntegers()) }

        val result = CollectionGenerators.mapOf(keyGenerator, valueGenerator)
        assertThat(result, notNullValue())
        assertThat(result.isEmpty(), `is`(false))

        //'At least', in case duplicate entries were generated
        verify(keyGenerator, atLeast(result.size)).get()
        verify(valueGenerator, atLeast(result.size)).get()
    }

    @Test
    fun testFromList()
    {
        println("testFromList")

        val list = ArrayList<String>()

        val size = one(integers(1, 100))
        for (i in 0..size - 1)
        {
            list.add(one(hexadecimalString(15)))
        }

        val instance = CollectionGenerators.fromList(list)
        assertThat(instance, notNullValue())

        doInLoop {
            val value = instance.get()
            assertThat(list.contains(value), `is`(true))
        }
    }


    @Test
    fun testListOfEdgeCases()
    {
        println("testListOfEdgeCases")

        val badSize = one(negativeIntegers())
        assertThrows { CollectionGenerators.listOf(StringGenerators.uuids(), badSize) }
                .isInstanceOf(IllegalArgumentException::class.java)

        val result = CollectionGenerators.listOf(StringGenerators.uuids, 0)
        assertThat(result, notNullValue())
        assertThat(result, `is`(empty<String>()))

    }
}
